salt/modules/nacl.py

```diff
@@ -130,91 +175,286 @@ def _get_config(**kwargs):
     '''
-    Return key
+    Return sk
     '''
     config = _get_config(**kwargs)
-    key = config['key']
-    keyfile = config['keyfile']
-    if not key and keyfile:
-        if not os.path.isfile(keyfile):
-            raise Exception('file not found: {0}'.format(keyfile))
-        with salt.utils.files.fopen(keyfile, 'rb') as keyf:
-            key = keyf.read()
+    key = config['sk']
+    sk_file = config['sk_file']
+    if not key and sk_file:
+        with salt.utils.fopen(sk_file, 'rb') as keyf:
+            key = str(keyf.read()).rstrip('\n')
     if key is None:
-        raise Exception('no key found')
-    key = str(key)
-    if rstrip_newline:
-        key = key.rstrip('\n')
-    return key
+        raise Exception('no key or sk_file found')
+    return base64.b64decode(key)
 
 
-def keygen(keyfile=None):
+def _get_pk(**kwargs):
+    '''
+    Return pk
     '''
-    Use libnacl to generate a private key
+    config = _get_config(**kwargs)
+    pubkey = config['pk']
+    pk_file = config['pk_file']
+    if not pubkey and pk_file:
+        with salt.utils.fopen(pk_file, 'rb') as keyf:
+            pubkey = str(keyf.read()).rstrip('\n')
+    if pubkey is None:
+        raise Exception('no pubkey or pk_file found')
+    pubkey = str(pubkey)
+    return base64.b64decode(pubkey)
+
+
+def keygen(sk_file=None, pk_file=None):
+    '''
+    Use libnacl to generate a keypair.
+
+    If no `sk_file` is defined return a keypair.
+
+    If only the `sk_file` is defined `pk_file` will use the same name with a postfix `.pub`.
+
+    When the `sk_file` is already existing, but `pk_file` is not. The `pk_file` will be generated
+    using the `sk_file`.
 
     CLI Examples:
 
     .. code-block:: bash
 
+        salt-call nacl.keygen
+        salt-call nacl.keygen sk_file=/etc/salt/pki/master/nacl
+        salt-call nacl.keygen sk_file=/etc/salt/pki/master/nacl pk_file=/etc/salt/pki/master/nacl.pub
         salt-call --local nacl.keygen
-        salt-call --local nacl.keygen keyfile=/root/.nacl
-        salt-call --local --out=newline_values_only nacl.keygen > /root/.nacl
     '''
-    b = libnacl.secret.SecretBox()
-    key = b.sk
-    key = base64.b64encode(key)
-    if keyfile:
-        if os.path.isfile(keyfile):
-            raise Exception('file already found: {0}'.format(keyfile))
-        with salt.utils.files.fopen(keyfile, 'w') as keyf:
-            keyf.write(key)
-            return 'saved: {0}'.format(keyfile)
-    return key
+    if sk_file is None:
+        kp = libnacl.public.SecretKey()
+        return {'sk': base64.b64encode(kp.sk), 'pk': base64.b64encode(kp.pk)}
+
+    if pk_file is None:
+        pk_file = '{0}.pub'.format(sk_file)
+
+    if sk_file and pk_file is None:
+        if not os.path.isfile(sk_file):
+            kp = libnacl.public.SecretKey()
+            with salt.utils.fopen(sk_file, 'w') as keyf:
+                keyf.write(base64.b64encode(kp.sk))
+            if not salt.utils.is_windows():
```

Comment:
@cachedout Maybe if it is Windows:
```
import salt.utils.win_functions
import salt.utils.win_dacl
cur_user = salt.utils.win_functions.get_current_user()
salt.utils.win_dacl.set_owner(sk_file, cur_user)
salt.utils.win_dacl.set_permissions(sk_file, cur_user, 'full_control', 'grant', reset_perms=True, protected=True)
```
That will make the user running the command the owner and remove all permissions to the file except for the current user. The current user will have full control permissions.